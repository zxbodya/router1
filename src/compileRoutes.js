import compileExpression from './expressions/compile.js';
import generatorFromExpression from './expressions/createGenerator.js';
import matcherFromExpression from './expressions/createMatcher.js';

function compileRoutes(routeDefs) {
  if (process.env.NODE_ENV !== 'production') {
    const usedNames = new Set();
    for (let i = 0, l = routeDefs.length; i < l; i++) {
      const routeDef = routeDefs[i];
      if (!routeDef.name) {
        throw new Error('routes should have name property');
      }
      if (!routeDef.handler) {
        throw new Error('routes should have handler property');
      }
      if (/[\/.]/.test(routeDef.name)) {
        throw new Error('route name should not contain slashes and dots');
      }
      if (usedNames.has(routeDef.name)) {
        throw new Error('route names should be uniq');
      }
      usedNames.add(routeDef.name);
    }
  }

  const rawRoutes = [];

  for (let i = 0, l = routeDefs.length; i < l; i++) {
    const routeDef = routeDefs[i];

    const urlParts = (routeDef.url || '').match(/^([^?]*)(?:\?(.*))?$/);

    const pathExpression = compileExpression(urlParts[1]);
    const searchParams = urlParts[2] ? urlParts[2].split('&') : [];

    // todo: warning if query param is overridden by search param
    // todo: warning if param is duplicated

    rawRoutes.push({
      name: routeDef.name,
      handler: routeDef.handler,
      matchPath: matcherFromExpression(pathExpression),
      generatePath: generatorFromExpression(pathExpression),
      searchParams: searchParams,
    });
  }
  return rawRoutes;
}

export default compileRoutes;
