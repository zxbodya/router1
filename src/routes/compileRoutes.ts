import { compile } from '../expressions/compile';
import { createGenerator } from '../expressions/createGenerator';
import { createMatcher } from '../expressions/createMatcher';

import { concat } from '../expressions/concat';
import { normalizeParams } from './normalizeParams';

import { Expresssion } from '../expressions/compile';
import { RouteParams } from '../Router';

export interface RouteDef<RouteHandler> {
  name?: string;
  url?: string;
  handler?: RouteHandler;
  handlers?: RouteHandler[];
  routes?: Array<RouteDef<RouteHandler>>;
}

interface RawRouteDef<RouteHandler> {
  names: Array<string | null | void>;
  handlers: RouteHandler[];
  pathExpressions: Expresssion[];
  searchParams: string[];
}

function parseRoutes<RouteHandler>(
  routeDefs: Array<RouteDef<RouteHandler>>
): Array<RawRouteDef<RouteHandler>> {
  const rawRoutes = [] as Array<RawRouteDef<RouteHandler>>;

  for (let i = 0, l = routeDefs.length; i < l; i += 1) {
    const routeDef = routeDefs[i];

    const urlParts = (routeDef.url || '').match(/^([^?]*)(?:\?(.*))?$/) as [
      string,
      string,
      string
    ];

    const pathExpression = compile(urlParts[1]);
    const searchParams = urlParts[2] ? urlParts[2].split('&') : [];

    if (routeDef.routes) {
      const nestedRoutes = parseRoutes(routeDef.routes);
      for (let ni = 0, nl = nestedRoutes.length; ni < nl; ni += 1) {
        const nestedRoute = nestedRoutes[ni];
        let handlers;
        if (routeDef.handlers) {
          handlers = [...routeDef.handlers, ...nestedRoute.handlers];
        } else {
          handlers = routeDef.handler
            ? [routeDef.handler, ...nestedRoute.handlers]
            : nestedRoute.handlers;
        }

        rawRoutes.push({
          names: [routeDef.name, ...nestedRoute.names],
          handlers,
          pathExpressions: [pathExpression, ...nestedRoute.pathExpressions],
          searchParams: [...searchParams, ...nestedRoute.searchParams],
        });
      }
    } else {
      rawRoutes.push({
        names: [routeDef.name],
        handlers:
          routeDef.handlers || (routeDef.handler ? [routeDef.handler] : []),
        pathExpressions: [pathExpression],
        searchParams,
      });
    }
  }
  return rawRoutes;
}

function getRouteName<RouteHandler>(routeDef: RawRouteDef<RouteHandler>) {
  return routeDef.names.filter(v => v).join('.');
}

export interface Route<RouteHandler> {
  name: string;
  handlers: RouteHandler[];
  matchPath: (path: string) => RouteParams | null;
  generatePath: (params: RouteParams) => string;
  searchParams: string[];
}

export function compileRoutes<RouteHandler>(
  routeDefs: Array<RouteDef<RouteHandler>>
): Array<Route<RouteHandler>> {
  const rawRoutes = parseRoutes(routeDefs);
  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production') {
    const usedNames = new Set();
    for (let i = 0, l = rawRoutes.length; i < l; i += 1) {
      const rawRoute = rawRoutes[i];
      const routeName = getRouteName(rawRoute);
      const pathExpression = concat(rawRoute.pathExpressions);

      // few routes with same name
      if (usedNames.has(routeName)) {
        console.warn(
          `"${routeName}" is dublicated, route names should be uniq`
        );
      }
      usedNames.add(routeName);

      // no handlers defined
      if (rawRoute.handlers.length === 0) {
        console.warn(`route "${routeName}" have no handlers`);
      }

      const { searchParams } = rawRoute;
      const pathParams = pathExpression[2];

      // if param is duplicated
      const pathParamsSet = new Set();
      for (let j = 0, sl = pathParams.length; j < sl; j += 1) {
        const paramName = pathParams[j];
        if (pathParamsSet.has(paramName)) {
          console.warn(
            `path param "${paramName}" in route "${routeName}" is duplicated`
          );
        }
        pathParamsSet.add(paramName);
      }

      // path param is overridden by search param
      for (let j = 0, sl = searchParams.length; j < sl; j += 1) {
        const paramName = searchParams[j];
        if (pathParamsSet.has(paramName)) {
          console.warn(
            `path param "${paramName}" in route "${routeName}" is overridden by search params`
          );
        }
      }
    }
  }
  return rawRoutes.map(rawRoute => {
    const routeName = getRouteName(rawRoute);
    const pathExpression = concat(rawRoute.pathExpressions);

    const matchPath = createMatcher(pathExpression);
    let generatePath = createGenerator(pathExpression);
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      const generatePathOriginal = generatePath;
      // test if matcher can match the result, and warn if not
      generatePath = params => {
        const normalizedParams = normalizeParams(rawRoute.searchParams, params);
        const path = generatePathOriginal(params);
        const match = matchPath(path);
        if (!match) {
          console.warn(
            `path "${path}" generated for route "${routeName}" with params ${JSON.stringify(
              params
            )}, is not matched by same route`
          );
        } else if (
          Object.keys(match).reduce(
            (acc, k) => acc || normalizedParams[k] !== match[k],
            false
          )
        ) {
          console.warn(
            `path "${path}" generated for route "${routeName}" with params ${JSON.stringify(
              params
            )}, is matched by same route with different params ${JSON.stringify(
              match
            )}`
          );
        }
        return path;
      };
    }
    return {
      name: routeName,
      handlers: rawRoute.handlers,
      matchPath,
      generatePath,
      searchParams: rawRoute.searchParams,
    };
  });
}
