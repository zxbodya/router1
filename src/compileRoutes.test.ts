import { compileRoutes } from './compileRoutes';

describe('Router, compiling route collection', () => {
  it('compiles empty collection', () => {
    expect(compileRoutes([])).toEqual([]);
  });

  it('compiles flat route collection', () => {
    const compiled = compileRoutes([
      {
        name: 'route1',
        handler: {},
        url: 'aaa<param1>bbb',
      },
      {
        name: 'route2',
        handler: {},
        url: 'ccc<param2>ddd',
      },
      {
        name: 'route3',
        handler: {},
        url: 'ccc<param2>ddd?a',
      },
      {
        name: 'route4',
        handler: {},
        url: 'ccc<param2>ddd',
      },
      {
        name: 'route5',
        handler: {},
        url: 'ccc<param2>ddd?a&b',
      },
    ]);

    expect(compiled[0].name).toEqual('route1');
    expect(compiled[0].searchParams).toEqual([]);

    expect(compiled[1].name).toEqual('route2');
    expect(compiled[1].searchParams).toEqual([]);

    expect(compiled[2].name).toEqual('route3');
    expect(compiled[2].searchParams).toEqual(['a']);

    expect(compiled[3].name).toEqual('route4');
    expect(compiled[3].searchParams).toEqual([]);

    expect(compiled[4].name).toEqual('route5');
    expect(compiled[4].searchParams).toEqual(['a', 'b']);
  });

  it('compiles nested route collection', () => {
    const compiled = compileRoutes([
      {
        name: 'a',
        handler: {},
        url: 'aaa<param1>bbb',
      },
      {
        name: 'b',
        handler: {},
        url: 'ccc<param2>',
        routes: [
          {
            name: '',
            url: 'ddd',
            routes: [
              {
                name: 'd',
                handler: {},
                url: '?a',
              },
              {
                name: 'e',
                handler: {},
                url: '?a&b',
              },
            ],
          },
        ],
      },
    ]);

    expect(compiled[0].name).toEqual('a');
    expect(compiled[0].searchParams).toEqual([]);

    expect(compiled[1].name).toEqual('b.d');
    expect(compiled[1].searchParams).toEqual(['a']);

    expect(compiled[2].name).toEqual('b.e');
    expect(compiled[2].searchParams).toEqual(['a', 'b']);
  });

  describe('warnings', () => {
    const { warn } = console;
    let lastWarnMessage: any[];
    beforeAll(() => {
      // tslint:disable-next-line no-console
      console.warn = (...args: any[]) => {
        lastWarnMessage = args;
      };
    });
    afterAll(() => {
      // tslint:disable-next-line no-console
      console.warn = warn;
    });
    beforeEach(() => {
      lastWarnMessage = null;
    });
    it('when route handler is missing', () => {
      compileRoutes([{ name: 'aaa' }]);
      expect(lastWarnMessage).toMatchSnapshot();
    });
    it('when route names are not uniq', () => {
      compileRoutes([
        { name: 'aaa', handler: '1' },
        { name: 'aaa', handler: '2' },
      ]);
      expect(lastWarnMessage).toMatchSnapshot();
    });
    it('when path param is duplicated', () => {
      compileRoutes([{ name: 'aaa', url: '<a><a>' }]);
      expect(lastWarnMessage).toMatchSnapshot();
    });
    it('when path param is overridden by search', () => {
      compileRoutes([{ name: 'aaa', url: '<a>?a' }]);
      expect(lastWarnMessage).toMatchSnapshot();
    });
    it('generating not properly matched url', () => {
      const [{ generatePath }] = compileRoutes([
        { name: 'aaa', url: '<a:\\w+><b>' },
      ]);
      generatePath({ a: '1', b: '2' });
      expect(lastWarnMessage).toMatchSnapshot();
    });
    it('generating not matched url', () => {
      const [{ generatePath }] = compileRoutes([{ name: 'aaa', url: '<a:1>' }]);
      generatePath({ a: '2' });
      expect(lastWarnMessage).toMatchSnapshot();
    });
  });
  it('allows to pass array of handlers', () => {
    const compiled = compileRoutes([
      {
        name: 'a1',
        handlers: ['0', '1'],
      },
    ]);

    expect(compiled[0].name).toEqual('a1');
    expect(compiled[0].handlers).toEqual(['0', '1']);
  });
  it('allows to pass array of handlers, prepends it with singular handler', () => {
    const compiled = compileRoutes([
      {
        name: 'a2',
        handler: '2',
        routes: [{ name: 'a1', handlers: ['0', '1'] }],
      },
    ]);

    expect(compiled[0].name).toEqual('a2.a1');
    expect(compiled[0].handlers).toEqual(['2', '0', '1']);
  });

  it('allows to pass array of handlers, appends it with singular handler', () => {
    const compiled = compileRoutes([
      {
        name: 'a3',
        handlers: ['0', '1'],
        routes: [{ name: 'a1', handler: '2' }],
      },
    ]);

    expect(compiled[0].name).toEqual('a3.a1');
    expect(compiled[0].handlers).toEqual(['0', '1', '2']);
  });

  it('allows to pass array of handlers, appends it with handlers array', () => {
    const compiled = compileRoutes([
      {
        name: 'a3',
        handlers: ['0', '1'],
        routes: [{ name: 'a1', handlers: ['0', '1'] }],
      },
    ]);

    expect(compiled[0].name).toEqual('a3.a1');
    expect(compiled[0].handlers).toEqual(['0', '1', '0', '1']);
  });
});
