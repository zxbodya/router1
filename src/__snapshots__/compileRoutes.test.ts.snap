// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router, compiling route collection warnings generating not matched url 1`] = `
Array [
  "path \\"2\\" generated for route \\"aaa\\" with params {\\"a\\":\\"2\\"}, is not matched by same route",
]
`;

exports[`Router, compiling route collection warnings generating not properly matched url 1`] = `
Array [
  "path \\"12\\" generated for route \\"aaa\\" with params {\\"a\\":\\"1\\",\\"b\\":\\"2\\"}, is matched by same route with different params {\\"a\\":\\"12\\",\\"b\\":\\"\\"}",
]
`;

exports[`Router, compiling route collection warnings when path param is duplicated 1`] = `
Array [
  "path param \\"a\\" in route \\"aaa\\" is duplicated",
]
`;

exports[`Router, compiling route collection warnings when path param is overridden by search 1`] = `
Array [
  "path param \\"a\\" in route \\"aaa\\" is overridden by search params",
]
`;

exports[`Router, compiling route collection warnings when route handler is missing 1`] = `
Array [
  "route \\"aaa\\" have no handlers",
]
`;

exports[`Router, compiling route collection warnings when route names are not uniq 1`] = `
Array [
  "\\"aaa\\" is dublicated, route names should be uniq",
]
`;
